### DISCUSSION TITLE
New XmpProfile for PNG

### DISCUSSION THREAD

**Discussion by KyleKolander** [dlemstra/Magick.NET#1603] [Help]
What is the best way to add XMP metadata to a PNG iimage?  

Regardless of whether I add to an existing XmpProfile, or create a new one, I get these types of warnings from exiftool.  If the image already had XMP metadata, exiftool could extract it fine... until I use Magick.NET to add XMP metadata.

Warning                         : [minor] Empty XMP
Warning                         : [minor] Text/EXIF chunk(s) found after PNG IDAT (may be ignored by some readers)

And it will no longer show the XMP metadata that it understood before this.

**Comment by dlemstra** [comment]
It looks like you are creating an empty xmp profile? If you give me something to reproduce this error message I might be able to give you an answer why this is happening.

**Reply by KyleKolander** [reply]
I have another example involving ACDSee XMP metadata, but let's start simple.

I attached the [PNG_XMP.zip](https://github.com/dlemstra/Magick.NET/files/14898246/PNG_XMP.zip) file with the following files:

- Iowa_original.png
- Iowa_exiftool_Rating.png
- Iowa_exiftool_Rating.xmp
- Iowa_Dopus_Rating.png
- Iowa_Dopus_Rating.xmp
- Iowa_Magick.NET_Rating.png
- Iowa_Magick.NET_Rating.xmp

I started out with the Iowa_original.png file.

Created the exiftool files:
```powershell
    cp Iowa_original.png Iowa_exiftool_Rating.png
    exiftool.exe -IFD0:Rating=4 -IFD0:RatingPercent=75 -XMP-xmp:Rating=4 -XMP-microsoft:RatingPercent=75 Iowa_exiftool_Rating.png
    exiftool.exe -xmp -b -w xmp Iowa_exiftool_Rating.png
```

Created the Directory Opus files:
```powershell
    cp Iowa_original.png Iowa_Dopus_Rating.png
    # manually set the star Rating using Directory Opus
    exiftool.exe -xmp -b -w xmp Iowa_Dopus_Rating.png
```

Created the Magic.NET files:
```powershell
    # ran the following program code to create Iowa_Magick.NET_Rating.png
    exiftool.exe -xmp -b -w xmp Iowa_Magick.NET_Rating.png
```

exiftool reports these warnings
- Directory Opus PNG:  `[minor] Fixed incorrect URI for xmlns:MicrosoftPhoto`
- Magick.NET PNG    :  `[minor] Empty XMP`

When I run the following command against all three of the Rating PNG files, I get the answer 4 back for both the exiftool and the Dopus files, but no answer from the Magick.NET file; instead, I get that warning message:
`Warning: [minor] Empty XMP`

What am I doing wrong with Magick.NET?  Is this the right/prescribed approach?  There are a lot of good examples, but I couldn't find any that involved creating an `XmpProfile` (i.e., processing a PNG without existing XMP metadata) and writing XMP metadata into the `MagickImage` instance.  In fact, I run into the same problem even if there is an existing `XmpProfile` and I add or modify a metadata value.


This is the simple program I wrote to test out Magick.NET XMP metadata handling for a PNG file.  The content string for the XDocument is just a copy/paste from Iowa_exiftool_Rating.xmp file.
```csharp
public class Program
{
    public static void Main()
    {
        const string input  = """P:\Web\downloads\Iowa_original.png""";
        const string output = """P:\Web\downloads\Iowa_Magick.NET_Rating.png""";

        using var image = new MagickImage(input);

        var (xmp, createdNewXmpProfile, xDocument) = GetXmpProfile(image);

        // logic to set/change the rating values...

        if (createdNewXmpProfile)
        {
            image.SetProfile(xmp);
        }

        image.Write(output);
    }

    public static (IXmpProfile, bool, XDocument) GetXmpProfile(MagickImage image)
    {
        var        createdNewXmpProfile = false;
        XDocument? xdoc;
        var        xmp = image.GetXmpProfile();
        if (xmp is null)
        {
            var content = """
                          <?xpacket begin='﻿' id='W5M0MpCehiHzreSzNTczkc9d'?>
                          <x:xmpmeta xmlns:x='adobe:ns:meta/' x:xmptk='Image::ExifTool 12.81'>
                          <rdf:RDF xmlns:rdf='http://www.w3.org/1999/02/22-rdf-syntax-ns#'>

                           <rdf:Description rdf:about=''
                            xmlns:MicrosoftPhoto='http://ns.microsoft.com/photo/1.0'>
                            <MicrosoftPhoto:Rating>75</MicrosoftPhoto:Rating>
                           </rdf:Description>

                           <rdf:Description rdf:about=''
                            xmlns:xmp='http://ns.adobe.com/xap/1.0/'>
                            <xmp:Rating>4</xmp:Rating>
                           </rdf:Description>
                          </rdf:RDF>
                          </x:xmpmeta>
                          <?xpacket end='r'?>
                          """;

            xdoc = XDocument.Parse(content);

            xmp                  = XmpProfile.FromXDocument(xdoc);
            createdNewXmpProfile = true;
        }
        else
        {
            xdoc = xmp.ToXDocument();
        }

        return (xmp, createdNewXmpProfile, xdoc);
    }
}
```

EDIT: Not that it should matter for a technical question like this, but the reason I'm doing this is because Directory Opus supports Rating PNG files, but Windows 11 File Explorer (and Properties dialog) does not.  I just reverse engineered what they are doing to accomplish this so I can programmatically do the same.

**Comment by StarGeekSpaceNerd** [comment]
[My hypothesis](https://exiftool.org/forum/index.php?topic=15918.msg85560#msg85560) on the exiftool forums.  There's a BOM (Byte Order Mark) at the beginning of the XMP block that I don't think should be there, though Phil will have to verify it.

There has been trouble before with Directory Opus and BOMs in the metadata on the exiftool forums.  There are several threads started by user mazeckenrode which have details.

**Reply by KyleKolander** [reply]
I was able to fix it thanks to @StarGeekSpaceNerd and Phil Harvey for helping to point me in the right direction.

I initially tried writing the XDocument to a MemoryStream and passing that to the XmpProfile constructor, but that resulted in exiftool complaining that the `XMP is double UTF-encoded`.  So I used a C# UTF8 suffix `u8`, which creates the string as a ReadOnlySpan<byte> and just passed that byte array to a different XmpProfile constructor.  Now the XMP data is identical to that which is created by the exiftool itself.

However, exiftool spits out this warning now `[minor] Text/EXIF chunk(s) found after PNG IDAT (may be ignored by some readers)`.  Magick.NET adds information to the PNG that was not in the original.

Before the IDAT chunks
```
  chunk cHRM at offset 0x00025, length 32
    White x = 0.3127 y = 0.329,  Red x = 0.64 y = 0.33
    Green x = 0.3 y = 0.6,  Blue x = 0.15 y = 0.06
  chunk bKGD at offset 0x00051, length 6
    red = 0x00ff, green = 0x00ff, blue = 0x00ff
  chunk zTXt at offset 0x00078, length 421, keyword: Raw profile type xmp
```
After the IDAT chunks, but before the IEND chunk, which it seems exiftool doesn't like:
```
  chunk tEXt at offset 0xf33d5, length 37, keyword: date:create
  chunk tEXt at offset 0xf3406, length 37, keyword: date:modify
  chunk tEXt at offset 0xf3437, length 40, keyword: date:timestamp
```

1. Why does Magick.NET do this?
2. Is there a way to tell Magick.NET to NOT do this?
3. Since exiftool complains about the tEXt chunks at the end, I'm guessing that is not per spec.  Can anyone confirm?
4. I'm still not sure whether my sample code is the "right" way to do this... It would be helpful if there was some documentation around this use case.

The updated code:
```csharp
public class Program
{
    public static void Main()
    {
        const string input  = """P:\Web\downloads\Iowa_original.png""";
        const string output = """P:\Web\downloads\Iowa_Magick.NET_Rating_NoXmlDeclaration.png""";

        using var image = new MagickImage(input);

        var (xmp, createdNewXmpProfile, xDocument) = GetXmpProfile(image);

        // logic to set/change the rating values...

        if (createdNewXmpProfile)
        {
            image.SetProfile(xmp);
        }

        image.Write(output);
    }

    public static (IXmpProfile, bool, XDocument) GetXmpProfile(MagickImage image)
    {
        var        createdNewXmpProfile = false;
        XDocument? xdoc;
        var        xmp = image.GetXmpProfile();
        if (xmp is null)
        {
            var content = """
                          <?xpacket begin='﻿' id='W5M0MpCehiHzreSzNTczkc9d'?>
                          <x:xmpmeta xmlns:x='adobe:ns:meta/' x:xmptk='Image::ExifTool 12.81'>
                          <rdf:RDF xmlns:rdf='http://www.w3.org/1999/02/22-rdf-syntax-ns#'>

                           <rdf:Description rdf:about=''
                            xmlns:MicrosoftPhoto='http://ns.microsoft.com/photo/1.0'>
                            <MicrosoftPhoto:Rating>75</MicrosoftPhoto:Rating>
                           </rdf:Description>

                           <rdf:Description rdf:about=''
                            xmlns:xmp='http://ns.adobe.com/xap/1.0/'>
                            <xmp:Rating>4</xmp:Rating>
                           </rdf:Description>
                          </rdf:RDF>
                          </x:xmpmeta>
                          <?xpacket end='r'?>
                          """u8;

            var contentBytes = content.ToArray();

            xmp                  = new XmpProfile(contentBytes);
            xdoc                 = xmp.ToXDocument();
            createdNewXmpProfile = true;
        }
        else
        {
            xdoc = xmp.ToXDocument();
        }

        return (xmp, createdNewXmpProfile, xdoc);
    }
}
```

**Comment by StarGeekSpaceNerd** [comment]
> Why does Magick.NET do this?

I believe it is allowed by the spec, so they are not in the wrong.  The reason exiftool flags it is because some programs do not read data in those locations.  See [this exiftool forums thread](https://exiftool.org/forum/index.php?topic=12657.msg68443#msg68443).

**Reply by dlemstra** [reply]
Writing the text chunks after the IDAT is something ImageMagick has done for many years. Moving it before the IDAT is probably not something that we will be going to change.
